src/term/string.rs

```diff
@@ -0,0 +1,399 @@
+    /// will not consider a Unicode codepoint to be contained in a string if
+    /// it exists as part of a larger extended grapheme cluster.
+    ///
+    /// The time complexity of this method is `O(self.len() * needle.len())`.
+    pub fn contains(&self, needle: &str) -> bool {
+        if needle.is_empty() {
+            true
+        } else {
+            use grapheme_cluster_preservation::SearchEvent;
+
+            grapheme_cluster_preservation::search(self, needle)
+                .any(|e| matches!(e, SearchEvent::Match { .. }))
+        }
+    }
+
+    /// Returns a new `NclString` replacing every occurence of `from` with `to`.
+    ///
+    /// This method has time complexity `O(self.len() * from.len())`.
+    pub fn replace(&self, from: &str, to: &str) -> NclString {
+        let mut result = String::with_capacity(self.capacity());
+
+        if from.is_empty() {
+            // If `from` is empty then we:
+            //   1. insert `to` at the beginning.
+            result.push_str(to);
+            //   2. insert `to` after each cluster.
+            self.graphemes(true)
+                .flat_map(|grapheme| [grapheme, to])
+                .for_each(|s| result.push_str(s))
+        } else {
+            use grapheme_cluster_preservation::SearchEvent;
+
+            grapheme_cluster_preservation::search(self, from).for_each(|e| match e {
+                // If we hit a match...
+                SearchEvent::Match { since_last_match } => {
+                    // ...we write everything we've seen since the last match...
+                    result.push_str(since_last_match);
+                    // ...and then we replace the matched `from` with `to`.
+                    result.push_str(to);
+                }
+                // We also need to write anything remaining after the last match.
+                SearchEvent::LastNonMatch { non_match } => result.push_str(non_match),
+            });
+        }
+
+        result.into()
+    }
+
+    /// Returns the substring of `self` between `start` and `end`.
+    ///
+    /// Returns an error if:
+    ///    - either start or end is not a fraction
+    ///    - start < 0
+    ///    - start > end
+    ///
+    /// The time complexity of this method is `O(self.len())`.
+    pub fn substring(&self, start: &Rational, end: &Rational) -> Result<NclString, SubstringError> {
+        let Ok(start_usize) = usize::try_from(start) else {
+            return Err(SubstringError::NonIntStart(start.clone()));
+        };
+
+        let Ok(end_usize) = usize::try_from(end) else {
+            return Err(SubstringError::NonIntEnd(end.clone()));
+        };
+
+        // We need to know the length of the grapheme cluster iterator,
+        // and we have to walk the graphemes to do that, so we might
+        // as well build a Vec to make slicing out the substring easier.
+        let graphemes: Vec<_> = self.graphemes(true).collect();
```

Comment:
I would never have thought to do this! Great suggestion.

Two small caveats to this though. One is that we've replaced the allocation with an extra iteration of the length of the substring at the end, because we need to walk its grapheme clusters to figure out whether it was actually the length we wanted. The other is that we need to walk the entire length of the string in order to produce error messages. 

I think optimising for the happy path, and for shorter substring lengths, makes sense though, so I've implemented your suggestion.